<h3>To predict if the Breast Cancer is 'Benign' Or 'Malignant', using given Dataset.</h3>
To Build a Classification Model to produce Accurate Prediction on the Data Entered.


In [1]:
import numpy as np
import pandas as pd

# Loading data

In [3]:
df = pd.read_csv("breast_cancer.csv")
df

Sample code number  Clump Thickness  Uniformity of Cell Size  \
0               1000025                5                        1   
1               1002945                5                        4   
2               1015425                3                        1   
3               1016277                6                        8   
4               1017023                4                        1   
..                  ...              ...                      ...   
678              776715                3                        1   
679              841769                2                        1   
680              888820                5                       10   
681              897471                4                        8   
682              897471                4                        8   

     Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
0                           1                  1                            2   
1                           4                  5                            7   
2                           1                  1                            2   
3                           8                  1                            3   
4                           1                  3                            2   
..                        ...                ...                          ...   
678                         1                  1                            3   
679                         1                  1                            2   
680                        10                  3                            7   
681                         6                  4                            3   
682                         8                  5                            4   

     Bare Nuclei  Bland Chromatin  Normal Nucleoli  Mitoses  Class  
0              1                3                1        1      2  
1             10                3                2        1      2  
2              2                3                1        1      2  
3              4                3                7        1      2  
4              1                3                1        1      2  
..           ...              ...              ...      ...    ...  
678            2                1                1        1      2  
679            1                1                1        1      2  
680            3                8               10        2      4  
681            4               10                6        1      4  
682            5               10                4        1      4  

[683 rows x 11 columns]

In [4]:
df["Class"].unique()

array([2, 4], dtype=int64)

In [5]:
x=df.iloc[:,1:-1].values
y=df.iloc[:,-1].values.reshape(-1,1)

# Splitting Data into Test and Train set

In [7]:
from sklearn.model_selection import train_test_split,cross_val_score
x_train,x_test,y_train,y_test=train_test_split(x,y,train_size=0.8,random_state=42)


In [8]:
from sklearn.preprocessing import StandardScaler
ss=StandardScaler()
x_train_s=ss.fit_transform(x_train)
x_test_s=ss.transform(x_test)

# Applying Logistic Regression Cassification

In [9]:
from sklearn.linear_model import LogisticRegression
classi=LogisticRegression(random_state=0)
classi.fit(x_train_s,y_train)
y_hat=classi.predict(x_test_s)

f:\Anaconda\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


# Fiding Accuracy on Test Set

In [10]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_hat)

0.9562043795620438

# K Folds on the test set (k=10)

In [11]:
k_fold_prediction=cross_val_score(estimator=classi,X=x_train_s,y=y_train.ravel(),cv=10)
print("Average Accuracy of our Classification Model is : ",k_fold_prediction.mean()*100)
print("Standard Deviation of our Classification Model is : ",k_fold_prediction.std()*100)

Average Accuracy of our Classification Model is :  96.70707070707071
Standard Deviation of our Classification Model is :  2.129841886655531


## Thus Our Logistic Regression Model has Accuracy of 96.7%

# K Nearest Neighbour

In [16]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5,metric="minkowski",p=2)
knn.fit(x_train_s,y_train.ravel())
y_pred=knn.predict(x_test_s)
print(accuracy_score(y_test,y_pred)*100)

96.35036496350365


# Niave Bayes

In [19]:
from sklearn.naive_bayes import GaussianNB
gn = GaussianNB()
gn.fit(x_train_s,y_train.ravel())
y_pred=gn.predict(x_test_s)
accuracy_score(y_test,y_pred)*100

95.62043795620438

# Decision Tree

In [26]:
from sklearn.tree import DecisionTreeClassifier
dtc = DecisionTreeClassifier(random_state=4,criterion="entropy")
dtc.fit(x_train_s,y_train)
y_pred=dtc.predict(x_test_s)
accuracy_score(y_test,y_pred)

0.9708029197080292

# Random forest

In [29]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=100,random_state=42)
rf.fit(x_train_s,y_train.ravel())
y_pred=rf.predict(x_test_s)
accuracy_score(y_test,y_pred)*100

95.62043795620438

# Deep Learning ANN model

In [30]:
import tensorflow as tf

In [43]:
y_train[y_train==2]=1
y_train[y_train==4]=0
y_test[y_test==2]=1
y_test[y_test==4]=0

In [42]:

ann = tf.keras.models.Sequential()
ann.add(tf.keras.layers.Dense(units=10,activation="relu"))
ann.add(tf.keras.layers.Dense(units=9,activation="relu"))
ann.add(tf.keras.layers.Dense(units=1,activation="sigmoid"))
ann.compile(optimizer="adam",loss="binary_crossentropy",metrics=["accuracy"])
ann.fit(x_train_s,y_train,batch_size=21,epochs=50)


Epoch 1/50
26/26 [==============================] - 1s 2ms/step - loss: 0.5107 - accuracy: 0.7436
Epoch 2/50
26/26 [==============================] - 0s 2ms/step - loss: 0.3719 - accuracy: 0.8626
Epoch 3/50
26/26 [==============================] - 0s 2ms/step - loss: 0.2803 - accuracy: 0.9286
Epoch 4/50
26/26 [==============================] - 0s 2ms/step - loss: 0.2179 - accuracy: 0.9597
Epoch 5/50
26/26 [==============================] - 0s 3ms/step - loss: 0.1747 - accuracy: 0.9707
Epoch 6/50
26/26 [==============================] - 0s 3ms/step - loss: 0.1444 - accuracy: 0.9707
Epoch 7/50
26/26 [==============================] - 0s 2ms/step - loss: 0.1236 - accuracy: 0.9670
Epoch 8/50
26/26 [==============================] - 0s 3ms/step - loss: 0.1089 - accuracy: 0.9670
Epoch 9/50
26/26 [==============================] - 0s 2ms/step - loss: 0.0996 - accuracy: 0.9689
Epoch 10/50
26/26 [==============================] - 0s 2ms/step - loss: 0.0929 - accuracy: 0.9689
Epoch 11/50
26/26 [

In [44]:
y_pred=ann.predict(x_test_s)
y_pred[y_pred<0.5]=0
y_pred[y_pred>=0.5]=1
accuracy_score(y_test,y_pred)*100


5/5 [==============================] - 0s 2ms/step


97.8102189781022

# XGBoost 

In [45]:
from xgboost import XGBClassifier
xgb = XGBClassifier()
xgb.fit(x_train_s,y_train)
y_pred = xgb.predict(x_test_s)
accuracy_score(y_test,y_pred)*100

95.62043795620438

# catBoost

In [46]:
from catboost import CatBoostClassifier
cb = CatBoostClassifier()
cb.fit(x_train_s,y_train)
y_pred = cb.predict(x_test_s)
accuracy_score(y_test,y_pred)*100

Learning rate set to 0.007956
0:	learn: 0.6782997	total: 163ms	remaining: 2m 43s
1:	learn: 0.6650630	total: 165ms	remaining: 1m 22s
2:	learn: 0.6491090	total: 166ms	remaining: 55.2s
3:	learn: 0.6325871	total: 167ms	remaining: 41.6s
4:	learn: 0.6194898	total: 168ms	remaining: 33.5s
5:	learn: 0.6031401	total: 170ms	remaining: 28.2s
6:	learn: 0.5898657	total: 171ms	remaining: 24.2s
7:	learn: 0.5749412	total: 172ms	remaining: 21.3s
8:	learn: 0.5605991	total: 172ms	remaining: 19s
9:	learn: 0.5460065	total: 173ms	remaining: 17.2s
10:	learn: 0.5334847	total: 174ms	remaining: 15.7s
11:	learn: 0.5227791	total: 175ms	remaining: 14.4s
12:	learn: 0.5116543	total: 176ms	remaining: 13.4s
13:	learn: 0.5026265	total: 177ms	remaining: 12.5s
14:	learn: 0.4919171	total: 179ms	remaining: 11.7s
15:	learn: 0.4859628	total: 179ms	remaining: 11s
16:	learn: 0.4760204	total: 180ms	remaining: 10.4s
17:	learn: 0.4635698	total: 181ms	remaining: 9.88s
18:	learn: 0.4531432	total: 182ms	remaining: 9.4s
19:	learn: 0.4

96.35036496350365

<h2><b><i>Thus we got 97.8% accuracy with Deep Learning Neural Network Model